deep_qa/layers/tuple_matchers/threshold_tuple_matcher.py

```diff
@@ -0,0 +1,236 @@
+        # Add the weights for the final layer.
+        self.score_layer = self.add_weight(shape=(self.hidden_layer_width, 1),
+                                           initializer=initializations.get(self.hidden_layer_init),
+                                           name='%s_score' % self.name)
+
+    def get_output_shape_for(self, input_shapes):
+        # pylint: disable=unused-argument
+        return (input_shapes[0][0], 1)
+
+    @overrides
+    def compute_mask(self, input, input_mask=None):  # pylint: disable=unused-argument,redefined-builtin
+        # Here, we want to see if either of the inputs is all padding (i.e. the mask would be all 0s).
+        # If so, then the whole tuple_match should be masked, so we would return a 0, otherwise we
+        # return a 1.  As such, the shape of the returned mask is (batch size, 1).
+        if input_mask == [None, None]:
+            return None
+        # Each of the two masks in input_mask are of shape: (batch size, num_slots, num_slot_words)
+        mask1, mask2 = input_mask
+        return K.any(mask1, axis=-1, keepdims=True) * K.any(mask2, axis=-1, keepdims=True)
+
+    def get_output_mask_shape_for(self, input_shape):  # pylint: disable=no-self-use
+        # input_shape is [(batch_size, num_slots, num_slot_words_t1, embedding_dim),
+        # (batch_size, num_slots, num_slot_words_t2, embedding_dim)]
+        mask_shape = (input_shape[0][0], 1)
+        return mask_shape
+
+    def call(self, x, mask=None):
+        tuple1_input, tuple2_input = x    # tuple1 shape: (batch size, num_slots, num_slot_words_t1, embedding_dim)
+                                          # tuple2 shape: (batch size, num_slots, num_slot_words_t2, embedding_dim)
+        # Check that the tuples have the same number of slots.
+        assert K.int_shape(tuple1_input)[1] == K.int_shape(tuple2_input)[1]
+
+        # Expand tuple1 to shape: (batch size, num_slots, num_slot_words_t1, num_slot_words_t2, embedding_dim)
+        # First expand to shape: (batch size, num_slots, num_slot_words_tuple1, 1, embedding_dim)
+        expanded_tuple1 = K.expand_dims(tuple1_input, 3)
+        # Tile to desired dimensions.
+        tiled_tuple1 = K.repeat_elements(expanded_tuple1, K.int_shape(tuple2_input)[2], axis=3)
+
+        # Expand tuple2 to shape: (batch size, num_slots, num_slot_words_t1, num_slot_words_t2, embedding_dim)
+        # First expand to shape: (batch size, num_slots, 1, num_slot_words_tuple2, embedding_dim)
+        expanded_tuple2 = K.expand_dims(tuple2_input, 2)
+        # Tile to desired dimensions.
+        tiled_tuple2 = K.repeat_elements(expanded_tuple2, K.int_shape(tuple1_input)[2], axis=2)
+
+        # Generate the similarity scores of each of the word pairs.
+        # shape: (batch size, num_slots, num_slot_words_tuple1, num_slot_words_tuple2)
+        tuple_word_similarities = self.similarity_function.compute_similarity(tiled_tuple1, tiled_tuple2)
```

Comment:
One thing you can try here is using `K.tile` instead of `K.repeat_elements`. The implementation of `K.repeat_elements` in the tensorflow backend of keras is very naive (to the point of being wrong): https://github.com/fchollet/keras/blob/master/keras/backend/tensorflow_backend.py#L1541
It effectively makes K copies of the thing you want to copy in memory and then concatenates them together, because it doesn't use `tf.tile`. If you use `K.tile`, I think tensorflow will just keep K pointers to a single tensor, so this might help.